src/geom/SphPoint.cc

```diff
@@ -0,0 +1,244 @@
+ * the GNU General Public License along with this program. If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include <string>
+#include <sstream>
+
+#include <cmath>
+
+#include "Eigen/Geometry"
+
+#include "lsst/afw/geom/SphPoint.h"
+#include "lsst/pex/exceptions.h"
+
+namespace pexExcept = lsst::pex::exceptions;
+using namespace std;
+
+namespace lsst {
+namespace afw {
+namespace geom {
+    /**
+     * Wraps an angle to the interval [0, 2&pi;).
+     *
+     * @param angle the angle to be wrapped
+     * @return an angle equivalent to @c angle, but forced into the
+     *         interval [0, 2&pi;). May be the same object as @c angle if
+     *         no wrapping was required.
+     */
+    Angle wrap(Angle const & angle) {
+        Angle copy(angle);
+        copy.wrap();
+        return copy;
+    }
+
+    /**
+     * Angular distance between two points using the Haversine formula.
+     *
+     * Besides the differences between the two coordinates, we also input the
+     * cosine of the two declinations, so as to be thrifty with the use of
+     * trigonometric functions.
+     *
+     * @param deltaLon Difference between longitudes.
+     * @param deltaLat Difference between latitudes. Must be in the same
+     *                 sense as @c deltaLon.
+     * @param cosLat1, cosLat2 Cosines of the two points' latitudes.
+     * @return the distance between the two points
+     */
+    Angle haversine(Angle const & deltaLon,
+                    Angle const & deltaLat,
+                    double cosLat1,
+                    double cosLat2
+    ) {
+        double const sinLat = sin(deltaLat/2.0);
+        double const sinLon = sin(deltaLon/2.0);
+        double const havDDelta = sinLat * sinLat;
+        double const havDAlpha = sinLon * sinLon;
+        double const havD = havDDelta + cosLat1 * cosLat2 * havDAlpha;
+        double const sinDHalf = sqrt(havD);
+        return (2.0 * asin(sinDHalf)) * radians;
+    }
+
+    SphPoint::SphPoint(Angle const & longitude, Angle const & latitude) :
+                    _longitude(wrap(longitude).asRadians()),
+                    _latitude(latitude.asRadians()) {
+        if (!isfinite(_latitude) || !isfinite(_longitude)) {
+            throw pexExcept::InvalidParameterError("NaN detected in constructing (" +
+                                                   to_string(longitude.asDegrees()) +
+                                                   ", " + to_string(latitude.asDegrees()) +
+                                                   ")");
+        }
+        // For some reason abs(double) is ignored
+        if (fabs(_latitude) > HALFPI) {
+            throw pexExcept::OutOfRangeError("Angle " + to_string(latitude.asDegrees()) +
+                " is not a valid latitude.");
+        }
+    }
+
+    SphPoint::SphPoint(Point3D const & vector) {
+        double const norm = sqrt(vector.distanceSquared(Point3D()));
```

Comment:
@r-owen What's the preferred idiom for doing this in Python? It appears `Point::asEigen()` doesn't have a corresponding method.